### DISCUSSION TITLE
How to continue the span with TraceID and SpanID when using Auto-Instrumentation?

### DISCUSSION THREAD

**Discussion by GokulShrikanth** [open-telemetry/opentelemetry-java-instrumentation#3114] [Q&A] [answered]
Hi, I am trying to trace through multiple microservices using Auto-instrumentation for java, But since I am using NATS to communicate to the services, The Trace Continuation from Service A to Service B is broken, Is there a way to resolve this? In the case of manual instrumentation, We can send the context through. Is there a way to achieve this with auto-instrumentation also?

**Comment by jkwatson** [comment]
@GokulShrikanth since this is about auto-instrumentation, I'm going to move this discussion over to the instrumentation repository.

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @GokulShrikanth,

By NATS you mean https://nats.io/ ?
It looks like we don't have instrumentation for this messaging software, so unfortunately the context does not get propagated.

**Reply by GokulShrikanth** [reply]
Oh...I was able to solve it by propagating context through headers

**Reply by GokulShrikanth** [reply]
Is there a way to set automatically generated spans to a certain context?

**Reply by mateuszrzeszutek** [reply]
You mean, set the NATS headers? You can implement the [`TextMapSetter`](https://github.com/open-telemetry/opentelemetry-java/blob/main/context/src/main/java/io/opentelemetry/context/propagation/TextMapSetter.java) interface and inject the current context manually; something similar to this:

```java
class NatsMessageHeaderSetter implements TextMapSetter<Message> {
  void set(@Nullable Message carrier, String key, String value) {
    carrier.headers().put(key, value);
  }
}

GlobalOpenTelemetry.get()
  .getContextPropagators()
  .getTextMapPropagator()
  .inject(Context.current(), message, new NatsMessageHeaderSetter());
```

You'll need similar code with `TextMapGetter` & `extract()` on the receiving side to get the span context across.
